File: Algorithms/Sliding-Window/0005_longest_palindrome_substring.py
#sliding_window

class Solution:
    def longestPalindrome(self, s: str) -> str:
        if not s: 
            return ""

        def expand_around_center(left: int, right: int):
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return right - left - 1 #subtract 1 to compensate 
            
        start = 0
        end = 0

        for i in range(len(s)):
            odd_length = expand_around_center(i, i)
            even_length = expand_around_center(i, i + 1)
            curr_max = max(odd_length, even_length)

            if curr_max > (end - start):
                start = i - (curr_max - 1) // 2 #divide by 2 as it is a palindrome 
                end = i + curr_max // 2

        return s[start : end + 1]

def test_palindrome():
    sol = Solution()
    
    # Case 1: Standard odd-length
    assert sol.longestPalindrome("babad") in ["bab", "aba"]
    
    # Case 2: Standard even-length
    assert sol.longestPalindrome("cbbd") == "bb"
    
    # Case 3: Single character
    assert sol.longestPalindrome("a") == "a"
    
    # Case 4: All same characters
    assert sol.longestPalindrome("aaaaa") == "aaaaa"
    
    # Case 5: No palindrome longer than 1
    assert len(sol.longestPalindrome("ac")) == 1

    print("✅ All Python test cases passed!")

if __name__ == "__main__":
    test_palindrome()

File: Algorithms/Sliding-Window/0005_longest_palindrome_substring.c
// Practice for: 0005_longest_palindrome_substring.py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int expand (char *s, int left, int right, int n){
    while (left >= 0 && right < n && s[left] == s[right]){
        left--;
        right++;       
    }
    return right - left - 1;
}

char* longestPalindrome(char* s){
    int n = strlen(s);
    if (n == 0){
        return s;
    }
    int start = 0;
    int max_len = 0;

    for (int i = 0; i < n; i++){
        int odd_length = expand(s,i,i,n);
        int even_length = expand(s,i,i+1,n);
        int curr_len = (odd_length > even_length) ? odd_length : even_length;

        if (curr_len > max_len){
            start = i - (curr_len -1) / 2;
            max_len = curr_len;
        }
    }
    char* result = calloc((max_len +1),sizeof(char));
    for (int i = 0; i < max_len; i++){
        result[i] = s[start + i];

    }
    return result;
}

int main() {
    char* inputs[] = {"babad", "cbbd", "a", "aaaaa", "ac"};
    char* expected[] = {"bab", "bb", "a", "aaaaa", "a"}; // "bab" or "aba" are both fine
    int num_tests = 5;

    printf("--- Testing Longest Palindromic Substring (C) ---\n");

    for (int i = 0; i < num_tests; i++) {
        char* result = longestPalindrome(inputs[i]);
        
        printf("Input: %s | Result: %s ", inputs[i], result);
        
        // Basic check (since babad can be bab or aba, we check length)
        if (strlen(result) >= strlen(expected[i])) {
            printf("✅\n");
        } else {
            printf("❌ (Expected length %lu)\n", strlen(expected[i]));
        }

        // IMPORTANT: Free the memory allocated by malloc/calloc!
        free(result);
    }

    return 0;
}
